pkg/doc/jose/jwe.go

```diff
@@ -334,28 +334,38 @@ func deserializeAndDecodeHeaders(rawJWE *rawJSONWebEncryption) (*Headers, *Heade
 	return &protectedHeaders, &unprotectedHeaders, nil
 }
 
-func deserializeRecipients(rawJWE *rawJSONWebEncryption) ([]*Recipient, error) {
-	var recipients []*Recipient
+func parseDeserializeRecipients(rawJWE *rawJSONWebEncryption) ([]*Recipient, error) {
+	if rawJWE.Recipients != nil {
+		var recipients []*Recipient
+
+		err := json.Unmarshal(rawJWE.Recipients, &recipients)
+		if err != nil {
+			return nil, err
+		}
+
+		return recipients, nil
+	}
 
 	// If there is no recipients field, then we must be deserializing JWE with the flattened syntax as defined in
 	// https://tools.ietf.org/html/rfc7516#section-7.2.2.
-	if rawJWE.Recipients == nil {
-		recipients = make([]*Recipient, 1)
-		recipients[0] = &Recipient{EncryptedKey: rawJWE.B64SingleRecipientEncKey}
-
-		if rawJWE.SingleRecipientHeader != nil {
-			err := json.Unmarshal(rawJWE.SingleRecipientHeader, &recipients[0].Header)
-			if err != nil {
-				return nil, err
-			}
-		}
-	} else {
-		err := json.Unmarshal(rawJWE.Recipients, &recipients)
+	recipient := &Recipient{EncryptedKey: rawJWE.B64SingleRecipientEncKey}
+
+	if rawJWE.SingleRecipientHeader != nil {
+		err := json.Unmarshal(rawJWE.SingleRecipientHeader, &recipient.Header)
 		if err != nil {
 			return nil, err
 		}
 	}
```

Comment:
Thinking about this change made me realize the original code was missing the check if both `recipients` and `SingleRecipientHeader` are empty in which case should return an error.

add: 
`else return fmt.Errorf("")` here

if recipients list was empty and SingleRecipientHeader is missing, then an error should be returned. Either we have a list of recipients or we have SingleRecipientHeader. Both cannot be empty.

Or even better, check `if rawJWE.B64SingleRecipientEncKey == "" ||  rawJWE.SingleRecipientHeader == "")` before new line 351 and return error if true. These should not be empty if `raw.Recipients` list is empty.

Reply:
@Baha-sk  this condition breaks tests (about 17 tests). I do not want to include it in this PR.

Reply:
We can add it later.